### PR TITLE
Add unusual fires report to master.

### PR DESCRIPTION
**Pull request by Anthony-Calderaro**

### CODE REVIEW

File: src/report/css/report.styl [hunk 1]
```diff
@@ -153,6 +153,21 @@ input.share-link-input {
   }
 }
 
+#unusualFiresOptions {
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/css/report.styl]: Removed

File: src/report/js/ReportBuilder.js [hunk 2]
```diff
@@ -2157,7 +2160,656 @@ define([
...
+         
+          const handleAs = { handleAs: 'json' };
+          const promiseUrls = [];
+          let sourceOfData = 'MODIS' || 'VIIRS';
+          const queryPrefix = 'https://production-api.globalforestwatch.org/query';
+          const stateQuerySuffix = `9b9e56fc-270e-486d-8db5-e0a839c9a1a9?sql=SELECT%20iso,%20adm1,%20adm2,%20week,%20year,%20alerts%20as%20count,%20area_ha,%20polyname%20FROM%20data%20WHERE%20iso%20=%20%27${countryCode}%27%20AND%20adm1%20=%201%20AND%20polyname%20=%20%27admin%27%20AND%20fire_type%20=%20%27${sourceOfData}%27`;
+          const countrySuffix = `ff289906-aa83-4a89-bba0-562edd8c16c6?sql=SELECT%20iso,%20adm1,%20adm2,%20week,%20year,%20alerts%20as%20count,%20area_ha,%20polyname%20FROM%20data%20WHERE%20iso%20=%20%27${countryCode}%27%20AND%20polyname%20=%20%27admin%27%20AND%20fire_type%20=%20%27${sourceOfData}%27`;
+          const subregionReport = window.reportOptions.aois;
+          const countryReport = window.reportOptions.country !== 'ALL';
+          const queryUrl = subregionReport ?  `${queryPrefix}/${stateQuerySuffix}` : countryReport ? `${queryPrefix}/${countrySuffix}` : null;
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: added `()`.

File: src/report/js/ReportBuilder.js [hunk 3]
```diff
@@ -2157,7 +2160,656 @@ define([
...
+          const currentMonth = new Date().getMonth();
+
+          // Calculate the current Week of the current year
+          const today = new Date();
+          const startDateOfCurrentYear = new Date(today.getFullYear(), 0, 0);
+          const diff = today - startDateOfCurrentYear;
+          const oneDay = 1000 * 60 * 60 * 24;
+          const day = Math.floor(diff / oneDay);
+          let currentWeek = 1;
+          for (let i = 1; i < day; i++) i % 7 === 0 ? currentWeek += 1 : null;
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Opened up.

File: src/report/js/ReportBuilder.js [hunk 4]
```diff
@@ -2157,7 +2160,656 @@ define([
...
+                  historicalAlerts: [],
+                  historicalAverage: 0,
+                  sd1: 0,
+                  sd2: 0,
+                  currentYearAlerts: 0
+                }
+                historicalDataByWeek.push(historicalWeekObject);
+              };
+
+              // historicalDataByWeek now contains 52 placeholderobjects, so we push an array of all historical alerts from that week to each from our query response.
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Yes. I fixed this typo.

File: src/report/js/ReportBuilder.js [hunk 5]
```diff
@@ -2157,7 +2160,656 @@ define([
...
+              threeMonthDataObject.windowSD1 = standardDeviationSeries.slice(0);
+              threeMonthDataObject.windowSD2 = standardDeviation2Series.slice(0);
+              threeMonthDataObject.windowMean = windowAverages.slice(0);
+              threeMonthDataObject.windowSDMinus1 = standardDeviationMinus1Series.slice(0);
+              threeMonthDataObject.windowSDMinus2 = standardDeviationMinus2Series.slice(0);
+              
+              // We repeat the process above for 6 months (26 weeks).
+              const isolated26Weeks = [];
+              if (currentWeek > 24) {
+                debugger;
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Removed

File: src/report/js/ReportBuilder.js [hunk 6]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+                // Update the categories based on whether it's 3, 6, or 12 months selected.
+                let selection = $(this).text();
+                rangeOfMonths = selection.includes('12') ? 12 : selection.includes('6') ? 6 : 3; 
+                currentYearToDate = categoriesArray.slice(0, (currentMonth + 1));
+                updatedCategoriesArray = categoriesArray.slice(currentMonth + 1);
+                currentYearToDate.forEach(index => updatedCategoriesArray.push(index));
+                updatedCategoriesArray = updatedCategoriesArray.slice(12 - rangeOfMonths);
+
+                // Update the series data based on whether it's 3, 6, or 12 months selected.
+                seriesData = selection.includes('12') ? twelveMonthDataObject.currentYearFires.slice(0) : selection.includes('6') ? sixMonthDataObject.currentYearFires.slice(0) : threeMonthDataObject.currentYearFires.slice(0);
```
Comment (christianhoward) [inline src/report/js/ReportBuilder.js]: `()` on all of them below

File: src/report/js/ReportBuilder.js [hunk 7]
```diff
@@ -2157,7 +2160,656 @@ define([
...
+
+              // Save data on our 12 month global object for future reference.
+              twelveMonthDataObject.currentYearFires = twelveMonthSeriesData.slice(0);
+              twelveMonthDataObject.windowSD1 = twelveMonthStandardDeviationSeries.slice(0);
+              twelveMonthDataObject.windowSD2 = twelveMonthStandardDeviation2Series.slice(0);
+              twelveMonthDataObject.windowMean = twelveMonthWindowAverages.slice(0);
+              twelveMonthDataObject.windowSDMinus1 = twelveMonthStandardDeviationMinus1Series.slice(0);
+              twelveMonthDataObject.windowSDMinus2 = twelveMonthStandardDeviationMinus2Series.slice(0);
+
+              // Calculate unusual fire counts
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: STARTED FROM THE BOTTOM NOW WE HERE!

File: src/report/js/ReportBuilder.js [hunk 8]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+                    adjustedIndex = rangeOfMonths === 3 ? adjustedIndex + 39 : rangeOfMonths === 6 ? adjustedIndex + 26 : adjustedIndex;
+
+                    const fires = this.point.y;
+                    const fireOrFires = fires === 1 ? 'Fire' : 'Fires';
+                    const sd2 = twelveMonthDataObject.windowSD2[adjustedIndex]['1'];
+                    const sd1 = twelveMonthDataObject.windowSD1[adjustedIndex]['1'];
+                    const sdMinus1 = twelveMonthDataObject.windowSDMinus1[adjustedIndex]['1'];
+                    const sdMinus2 = twelveMonthDataObject.windowSDMinus2[adjustedIndex]['1'];
+
+                    const usuality = fires  > sd2 ? 'Unusually High' : fires  > sd1 ? 'High' : (fires  < sd1 && fires > sdMinus1) ? 'Average' : fires  < sdMinus2 ? 'Unusually Low' : 'Low';
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Bop it!

File: src/report/js/ReportBuilder.js [hunk 9]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+            timeOptions.forEach(period => $('#unusualFiresOptions').append(`<ul class=${period === '3 months'  ? 'selected' : ''}>` + period + "</ul>"));
+
+            // On click of a time option, we highlight it and update the series accordingly
+            $('#unusualFiresOptions ul').click(function() {
+                $('#unusualFiresOptions ul').removeClass('selected');
+                $(this).addClass('selected');
+                
+                // Update the categories based on whether it's 3, 6, or 12 months selected.
+                let selection = $(this).text();
+                rangeOfMonths = selection.includes('12') ? 12 : selection.includes('6') ? 6 : 3;
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Added

File: src/report/js/ReportBuilder.js (hunk 6 above)
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Added

File: src/report/js/ReportBuilder.js [hunk 10]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+              twelveMonthDataObject.windowSDMinus2 = twelveMonthStandardDeviationMinus2Series.slice(0);
+
+              // Calculate unusual fire counts. We need isolate all of the current week and current year data from the historical data.
+              arrayToFindFiresCount = dataFromRequest.filter(weekOfData => weekOfData.week === currentWeek && weekOfData.year === currentYear);
+              unusualFiresCount = arrayToFindFiresCount.length > 0 ? arrayToFindFiresCount['0'].alerts : 0;
+              earliestYearOfData = currentYear;
+              dataFromRequest.forEach(week => week.year < earliestYearOfData ? earliestYearOfData = week.year : earliestYearOfData);
+
+              // Since the current month should be last, we need to slice and reorder the months based on the current month and the rangeOfMonths selected.
+              currentYearToDate = categoriesArray.slice(0, (currentMonth + 1));
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Updated

File: src/report/js/ReportBuilder.js [hunk 11]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+            $('#unusualFiresOptions ul').click(function() {
+                $('#unusualFiresOptions ul').removeClass('selected');
+                $(this).addClass('selected');
+                
+                // Update the categories based on whether it's 3, 6, or 12 months selected.
+                let selection = $(this).text();
+                rangeOfMonths = selection.includes('12') ? 12 : selection.includes('6') ? 6 : 3; 
+                currentYearToDate = categoriesArray.slice(0, (currentMonth + 1));
+                updatedCategoriesArray = categoriesArray.slice(currentMonth + 1);
+                currentYearToDate.forEach(index => updatedCategoriesArray.push(index));
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Done

**Comment by Anthony-Calderaro** [comment]
Hold on merging until we receive feedback from Eric re:

- [ ] The sharing / exporting layout. This should be made consistent with the other charts.
- [ ] Hiding the chart on Global Report isn't working properly.
- [ ] Chart layout on subregions with 0 or 1 static piece of data. In these cases, it's impossible to calculate a deviation or mean, so do we hide the chart, or display something else?

File: src/report/js/ReportBuilder.js [hunk 12]
```diff
@@ -2157,7 +2160,658 @@ define([
...
+              * "High/Low" means that total fires are beyond +/- 1 SD
+              * "Unusually High/Low" means that total fires are beyond +/- 2 SD
+            ***************************************************/
+
+            const stndrdDev2 = twelveMonthDataObject.windowSD2[currentMonth - 1]['1'];
+            const stndrdDev1 = twelveMonthDataObject.windowSD1[currentMonth - 1]['1'];
+            const stndrdDevMin1 = twelveMonthDataObject.windowSDMinus1[currentMonth - 1]['1'];
+            const stndrdDevMin2 = twelveMonthDataObject.windowSDMinus2[currentMonth - 1]['1'];
+            
+            let currentWeekUsuality = unusualFiresCount > stndrdDev2 ? 'Unusually High' : unusualFiresCount  > stndrdDev1 ? 'High' : (unusualFiresCount  < stndrdDev1 && unusualFiresCount > stndrdDevMin1) ? 'Average' : unusualFiresCount  < stndrdDevMin2 ? 'Unusually Low' : 'Low';
```
[the remark this answers is not in the document]
Comment (Anthony-Calderaro) [reply src/report/js/ReportBuilder.js]: Used `if, else ifs, else` instead, but it's more readable now.

**Comment by christianhoward** [comment]
👍

**Pull request merged**